### PR TITLE
UI-GRID added some api extensions, fixed some types

### PR DESCRIPTION
**Pull request by raysuelzer**
added core, pagination, selection to GridAPI.  Also fixed a few types that were incorrect, missing, or should have been nullable.

### CODE REVIEW

**Comment by vvakame** [comment]
*ui-grid/ui-grid.d.ts*

to author(@btesser). could you review this PR?
:+1: or :-1:?

check list

* [ ] pass the Travic-CI test?

**Comment by btesser** [comment]
sure.  I'll take a look

File: ui-grid/ui-grid-tests.ts [hunk 1]
```diff
@@ -72,7 +72,7 @@ columnDef.menuItems = [{
 columnDef.minWidth = 100;
 columnDef.name = 'MyColumn';
 columnDef.sort = {
-    direction: 0,
+    direction: 'ASC',
```
Comment (btesser) [inline ui-grid/ui-grid-tests.ts]: You should use uiGridConstants.ASC which is equal to a number

File: ui-grid/ui-grid.d.ts [hunk 2]
```diff
@@ -88,11 +88,11 @@ declare module uiGrid {
     }
     export interface IGridOptions {
         aggregationCalcThrottle?: number;
-        appScopeProvider?: ng.IScope;
+        appScopeProvider?: ng.IScope | Object;
```
Comment (btesser) [inline ui-grid/ui-grid.d.ts]: The object needs to extend ng.IScope

File: ui-grid/ui-grid.d.ts [hunk 3]
```diff
@@ -88,11 +88,11 @@ declare module uiGrid {
...
         columnDefs?: IColumnDef;
         columnFooterHeight?: number;
         columnVirtualizationThreshold?: number;
-        data?: Array<any>;
+        data?: Array<any> | string;
```
Comment (btesser) [inline ui-grid/ui-grid.d.ts]: Can you comment on this?

File: ui-grid/ui-grid.d.ts [hunk 4]
```diff
@@ -604,7 +664,7 @@ declare module uiGrid {
         leaveOpen?: boolean;
     }
     export interface ISortInfo {
-        direction?: number;
+        direction?: string;
```
Comment (btesser) [inline ui-grid/ui-grid.d.ts]: This should be uiGridConstants.ASC or uiGridConstants.DESC which are both numbers

**Comment by btesser** [comment]
Just a few minor issues.  Clean those up and it should be good

**Comment by raysuelzer** [comment]
UiGridConstants.ASC returns a string as far as i can tell. 

https://github.com/angular-ui/ui-grid/blob/master/src/js/core/constants.js#L65

I can clean this up tomorrow.  I also added a few more things.

File: ui-grid/ui-grid.d.ts (hunk 2 above)
Comment (raysuelzer) [inline ui-grid/ui-grid.d.ts]: Actually, it does not as the internals extend it to the scope. For example:

                appScopeProvider: {
                    onDoubleClick: () => {
                        $rootScope.$emit("setActiveTab", 'member');
                    }
                },

File: ui-grid/ui-grid.d.ts (hunk 3 above)
Comment (raysuelzer) [inline ui-grid/ui-grid.d.ts]: You can pass in the string value of scope.

For example:

```
$scope.searchResults = []

 $scope.gridOptions = {
data: 'searchResults'
}
```

File: ui-grid/ui-grid.d.ts (hunk 4 above)
Comment (raysuelzer) [inline ui-grid/ui-grid.d.ts]: @btesser  -- UiGridConstants.ASC returns a string as far as i can tell:

https://github.com/angular-ui/ui-grid/blob/master/src/js/core/constants.js#L65

Comment (btesser) [inline ui-grid/ui-grid.d.ts]: Wow I guess I overlooked that... my bad

File: ui-grid/ui-grid.d.ts (hunk 3 above)
Comment (btesser) [inline ui-grid/ui-grid.d.ts]: Ahh.  I wasn't familiar with that.  Is it documented?

File: ui-grid/ui-grid.d.ts (hunk 2 above)
Comment (btesser) [inline ui-grid/ui-grid.d.ts]: Interesting.  I guess I made a poor assumption here

**Comment by raysuelzer** [comment]
I'll make those changes soon.

**Comment by btesser** [comment]
Cool. looks good to me other than the style stuff

**Comment by raysuelzer** [comment]
Looks like a left a commit on my local machine at work. I'll push it up tomorrow.

**Comment by btesser** [comment]
LGTM

**Comment by raysuelzer** [comment]
@vvakame - This is ready.

**Comment by vvakame** [comment]
@btesser thanks!
@raysuelzer thanks mate! :+1:

**Pull request merged**